### PR TITLE
[JENKINS-12699] Temp files for file parameters never deleted

### PR DESCRIPTION
**Pull request by bphinz**
See:

[JENKINS-12699](https://issues.jenkins-ci.org/browse/JENKINS-12699)
[JENKINS-32215](https://issues.jenkins-ci.org/browse/JENKINS-32215).

### Proposed changelog entries

* [JENKINS-12699](https://issues.jenkins-ci.org/browse/JENKINS-12699): Jenkins creates temp file object when using _File Parameters_ and never deletes it, possibly causing server to run out of disk space

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
      * I don't know how to write a test for this change but we've had this in our production Jenkins instance since November and no negative side effects have been observed.  The temp files are deleted after merging this in.
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### CODE REVIEW

**Review by jvz — COMMENTED** [review]
Seems logical. Do you think you can add a test for this?

**Comment by bphinz** [comment]
I could try, but I've never written a test like this.  Are there examples somewhere that I might be able to leverage?  What should the test consist of, just verifying that the files get deleted?

**Comment by jvz** [comment]
Yes, just verify the file was deleted. It might be doable with mocks, but I'd assume that you can use `WebClient` in the test to upload a file to a form and things like that.

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Thanks for the pull request @bphinz ! Unfortunately it will not work as expected, because the file will not be consumed until the build is actually executed in https://github.com/jenkinsci/jenkins/blob/881dafe385f42bdd2568c7286e1dfac30bec6daa/core/src/main/java/hudson/model/FileParameterValue.java#L173 . 

It might be possible to have better implementation, which would automatically move temporary files to the build directory when the build starts. In such case the files would be rotated along with build directories. Or it could just remove files, because they are browsed from the workspace anyway in https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/FileParameterValue.java#L228 

Such logic likely requires additional metadata which would indicate that the file is actually a temporary one

**Comment by bphinz** [comment]
Thanks for responding.  I like the idea of moving them into the build directory at the start of the build.  I'll see if I can work through that this weekend.

**Comment by oleg-nenashev** [comment]
thanks!

**Comment by bphinz** [comment]
I verified manually that with this patch the tmp files get deleted, but I'm not really sure how to go about writing an automated test for this...  The problem only presents itself when slaves are present, so the test would need to ensure that.  Also, the test would need to watch java.io.tmpdir on the master to see if a tmp file was created and then deleted.  Any thoughts?

**Comment by bphinz** [comment]
Hi @oleg-nenashev.  Are you still reviewing this?

**Comment by oleg-nenashev** [comment]
Sorry, I missed the previous comment. Will add to my queue

**Comment by jsoref** [comment]
@bphinz: you have ` `  + `x` in your `[]` which breaks the checkbox feature of github, can you make sure you have exactly one of those two characters between `[` and `]`?

**Comment by batmat** [comment]
@bphinz thank you very much for this work. That's an interesting fix. 

I also feel like this ought to have an automated test, both for checking we do reproduce/see the issue, and we keep it fixed once done.

I would recommend you look at [`hudson.model.FileParameterValueTest`](https://github.com/jenkinsci/jenkins/blob/c3351d2e7c3edfee82b9470e9aa1168982296072/test/src/test/java/hudson/model/FileParameterValueTest.java#L51) and the tests around this. It seems like you would _just_ to find a way to 

1) (by removing your fix first locally) write a test that fails by checking there's no temporary file left
2) add your fix and see if that goes back to success.

If you feel like this is something you could enjoy playing with, then we're happy to help. If you don't feel like it, or have no time anymore, please just tell us and I think we will try to see if we can add that test ourselves.

Thanks again!

**Comment by bphinz** [comment]
Sorry, didn't mean to let this fall off the radar.  I will attempt to create a testcase based on @batmat's suggestions.

**Comment by jglick** [comment]
I am not sure it is worth trying to salvage `FileParameter`, which has so many architectural problems to begin with. [Discussion](https://issues.jenkins-ci.org/browse/JENKINS-27413?focusedCommentId=352030&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-352030)

**Comment by batmat** [comment]
hello @bphinz! Are you still intending to finish this? I still think this is a useful fix, so if you need on how to write a test in/for Jenkins, just ask. Thanks!

**Comment by alecharp** [comment]
@bphinz do you think you would have time to add a test-case on this? Thanks.

**Comment by bphinz** [comment]
I'm really sorry for the delay!  Yes, I still want to try to build the
testcase

**Comment by bphinz** [comment]
I have a test that I think will work.  I will commit the patch later today

**Comment by alecharp** [comment]
@bpaquet Thanks for the test case but could you please review the merge conflict and try to resolve it?

**Comment by oleg-nenashev** [comment]
Fixed the merge conflict

**Review by oleg-nenashev — COMMENTED** [review]
I am in favor of making it a default behavior, but it may cause regressions in some rebuild functionality. I think more comments from @jenkinsci/core team needed before merging.

Also note that `FileItem#delete` does not really propagate deletion issues. Not a big deal for temporary files

File: test/src/test/java/hudson/model/FileParameterValueTest.java
Comment (alecharp) [inline test/src/test/java/hudson/model/FileParameterValueTest.java]:
```suggestion
    }

```

**Review by alecharp — CHANGES_REQUESTED** [review]
Requires to close the test function but other than that :+1:

File: test/src/test/java/hudson/model/FileParameterValueTest.java
Comment (oleg-nenashev) [reply test/src/test/java/hudson/model/FileParameterValueTest.java]: Fixed it manually

**Review by oleg-nenashev — APPROVED** [review]
To reiterate on my previous feedback, I am +1 for this change
It would be nice to have a system property to allow opting out, but I think we can start from the current behavior in weekly

**Comment by jglick** [comment]
Anyone happen to know if file parameters worked with the `rebuild` plugin before and, if so, whether this breaks the combination? (An example of the “architectural problems” I mentioned.)

**Comment by bphinz** [comment]
I will check today but I believe that rebuilds always worked and should
continue to do so.  The file gets copied into the workspace and distributed
to the slaves from there.  What this patch addresses is the tmp file that
only ever gets created on the master but then doesn't get deleted after
being copied to the workspace.

**Comment by bphinz** [comment]
Slight correction - I just tested and rebuilds from the web UI require the
file to be uploaded again irregardless of this patch

**Review by jeffret-b — APPROVED** [review]
I'm in favor of this change. I haven't tried testing it, but the code looks fine and I'm in favor of the improvement.

**Review by jeffret-b — CHANGES_REQUESTED** [review]
Hmm ... I got curious about the CI failure and went ahead and downloaded the PR, built it and ran the tests. Unfortunately, the new test is failing on my machine. I'm not sure what's going on.

I'm still in favor of the improvement of the PR but something in the PR still needs a little work.

**Comment by varyvol** [comment]
@bphinz could you review the test failures that @jeffret-b mentioned? It's actually the test that you introduced and it seems to be failing when checking the file does not exist anymore.

File: test/src/test/java/hudson/model/FileParameterValueTest.java
Comment (varyvol) [inline test/src/test/java/hudson/model/FileParameterValueTest.java]: assertNull() probably easier to read (I'm not doing the suggestion directly because it might require the import).

**Comment by fcojfernandez** [comment]
Since there is no activity, we're proposing this PR for close

**Comment by bphinz** [comment]
No argument from me, I can't spin my wheels on this anymore.  Note that this issue is probably worthy of a CVE because it allows a remote user to fill up /tmp on the jenkins master, potentially creating a DoS condition (that's actually how I discovered it in the first place).

**Comment by alecharp** [comment]
If so, it would be way better not discussing it here but on a security ticket.

**Comment by MRamonLeon** [comment]
Closed as agreed

**Pull request closed without merging**
